operator/internal/manifests/storage/schema.go

```diff
@@ -0,0 +1,200 @@
+package storage
+
+import (
+	"sort"
+	"time"
+
+	"github.com/ViaQ/logerr/v2/kverrors"
+	lokiv1beta1 "github.com/grafana/loki/operator/api/v1beta1"
+)
+
+const (
+	// DateTimeFormat is the datetime string need to format the time.
+	DateTimeFormat = "2006-01-02"
+
+	// UpdateDelay is the time before a change to the storage schema
+	// is applied to a cluster. This will avoid scenarios in which
+	// the difference between UTC and local time zones will cause
+	// logs to be unavailable.
+	UpdateDelay = time.Hour * 24
+)
+
+type schemaConfig struct {
+	version       lokiv1beta1.ObjectStorageSchemaVersion
+	effectiveDate time.Time
+}
+
+// BuildSchemaConfigList creates a list of schemas to be used to configure
+// the storage schemas for the cluster. This method assumes that the following
+// validation has been done to the statuses and specs:
+//
+// 1. All EffectiveDate fields are able to be parsed
+// 2. All EffectiveDate fields are unique in their respective list
+//
+func BuildSchemaConfigList(
+	utcTime time.Time,
+	specs []lokiv1beta1.ObjectStorageSchemaSpec,
+	statuses []lokiv1beta1.StorageSchemaStatus,
+) ([]lokiv1beta1.ObjectStorageSchemaSpec, error) {
+	if len(specs) == 0 {
+		return nil, kverrors.New("Spec does not contain any schemas.")
+	}
+
+	specSchemas := specSchemaConfigList(specs)
+
+	if len(statuses) != 0 {
+		statusSchemas := statusSchemaConfigList(statuses)
+
+		// Schemas applied in the future can be ignored cause they have not
+		// taken effect yet. The delay is added to avoid edge cases where UTC
+		// and local time zones may cause interference.
+		//
+		// For example, say there is a future change which will occur on 6/02/22.
+		// If change occurs close to midnight on 6/01/22, there might be timing
+		// issues between the local and UTC. Thus, there is a chance of corruption.
+		// To avoid this, the change on 6/02/22 should be considered as already
+		// applied and an error should be sent back to change the date of the
+		// in-flight change.
```

Comment:
nit. to improve reability of comments I suggest to keep to the date format we parse, i.e. `2022-06-02`.